File: projects/13-Hopcroft/implementation/hopcroft.py
import csv
import os
from collections import deque, defaultdict

# A lightweight class to model a Deterministic Finite Automaton (DFA)
class DFA:
    def __init__(self, states, alphabet, transition_function, start_state, accept_states):
        # Initialize DFA components: state set, input alphabet, transition map,
        # designated start state, and set of accepting states.
        self.states = states
        self.alphabet = alphabet
        self.tf = transition_function
        self.start = start_state
        self.accept = set(accept_states)

    def remove_unreachable_states(self):
        # Identify all states reachable from the start via breadth-first search
        reachable = set()
        queue = deque([self.start])

        while queue:
            state = queue.popleft()
            if state not in reachable:
                reachable.add(state)
                # Follow each possible input symbol to find next states
                for c in self.alphabet:
                    target = self.tf.get((state, c))
                    if target and target not in reachable:
                        queue.append(target)

        # Discard unreachable states and prune accept set accordingly
        self.states = reachable
        self.accept &= reachable

        # Rebuild transition function to include only reachable transitions
        self.tf = {
            (s, c): t for (s, c), t in self.tf.items()
            if s in reachable and t in reachable
        }

    def make_total(self):
        # Ensure every state has an outgoing transition for each symbol
        trap = "__TRAP__"
        needs_trap = False

        # Fill missing transitions with a trap placeholder
        for state in self.states:
            for symbol in self.alphabet:
                if (state, symbol) not in self.tf:
                    self.tf[(state, symbol)] = trap
                    needs_trap = True

        # If any gap was filled, add the trap state itself and loop it
        if needs_trap:
            self.states.add(trap)
            for symbol in self.alphabet:
                self.tf[(trap, symbol)] = trap

    def minimize(self):
        # Step 1: Prune unreachable states and complete the machine
        self.remove_unreachable_states()
        self.make_total()

        # Step 2: Initialize partitions: accepting vs non-accepting
        P = [self.accept, self.states - self.accept]
        W = deque(P.copy())  # worklist of partitions to refine

        # Step 3: Refine partitions until stable
        while W:
            A = W.popleft()

            # For each symbol, find states transitioning into A
            for c in self.alphabet:
                X = {s for s in self.states if self.tf.get((s, c)) in A}

                new_P = []
                for Y in P:
                    # Split Y into those in X vs outside X
                    inter = X & Y
                    diff = Y - X

                    if inter and diff:
                        # Replace Y with its two parts
                        new_P.extend([inter, diff])

                        # Manage worklist updates for efficiency
                        if Y in W:
                            W.remove(Y)
                            W.extend([inter, diff])
                        else:
                            # Always split the smaller block next
                            W.append(inter if len(inter) <= len(diff) else diff)
                    else:
                        # No split needed, carry Y forward
                        new_P.append(Y)

                # Commit all splits at once to current partition list
                P = new_P

        # Step 4: Assign each block a numeric label for new states
        sorted_blocks = sorted([sorted(block) for block in P], key=lambda b: b[0])
        block_map = {frozenset(block): i for i, block in enumerate(sorted_blocks)}

        # Build reverse map: state -> block index
        representative = {
            state: block_map[frozenset(block)]
            for block in sorted_blocks for state in block
        }

        # Identify new DFA components based on blocks
        new_states = set(block_map.values())
        new_start = representative[self._find_block(self.start, P)]
        new_accept = {representative[s] for s in self.accept}
        new_tf = {}

        # For each block, pick any member to determine outgoing transitions
        for block in P:
            rep = next(iter(block))
            for c in self.alphabet:
                target = self.tf.get((rep, c))
                if target is not None:
                    tgt_block = block_map[frozenset(self._find_block(target, P))]
                    new_tf[(representative[rep], c)] = tgt_block

        # Return a fresh DFA object representing the minimized machine
        return DFA(new_states, self.alphabet, new_tf, new_start, new_accept)

    def _find_block(self, state, partition):
        # Locate which block in the partition contains the given state
        for block in partition:
            if state in block:
                return tuple(block)  # return a sample member of that block
        return None

# Helpers for CSV I/O and testing follow the same pattern

def parse_dfa_from_csv(filename):
    """
    Read a DFA description from CSV with columns:
    state, is_start, is_accept, <symbol1>, <symbol2>, ...
    """
    with open(filename, 'r') as f:
        reader = csv.DictReader(f)
        rows = list(reader)
        alphabet = reader.fieldnames[3:]

        states = set()
        tf = {}
        start_state = None
        accept_states = set()

        # Gather states, start/accept flags, and transition entries
        for row in rows:
            state = row['state']
            states.add(state)
            if row['is_start'] == '1':
                start_state = state
            if row['is_accept'] == '1':
                accept_states.add(state)

            for sym in alphabet:
                tf[(state, sym)] = row[sym]

    return DFA(states, set(alphabet), tf, start_state, accept_states)


def write_dfa_to_csv(dfa, filename):
    # Prepare rows by grouping transitions per state
    state_data = defaultdict(dict)
    for (state, symbol), target in dfa.tf.items():
        state_data[state][symbol] = target

    # Build a sorted list of rows for predictable ordering
    rows = []
    for state in sorted(dfa.states):
        row = {
            'state': state,
            'is_start': '1' if state == dfa.start else '0',
            'is_accept': '1' if state in dfa.accept else '0'
        }
        for sym in sorted(dfa.alphabet):
            row[sym] = state_data[state].get(sym, '')
        rows.append(row)

    # Write out to file with header plus transition columns
    fieldnames = ['state', 'is_start', 'is_accept'] + sorted(dfa.alphabet)
    with open(filename, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def compare_csv_files(file1, file2):
    # Load each CSV as a list of dicts and compare equality
    def read_csv_as_dict(file):
        with open(file, 'r') as f:
            return list(csv.DictReader(f))
    return read_csv_as_dict(file1) == read_csv_as_dict(file2)


def run_test(index):
    # Automate a test run given an index for input/output filenames
    base = 'io'
    input_file = os.path.join(base, f'sample.in.{index}')
    expected_output = os.path.join(base, f'sample.out.{index}')
    generated_output = os.path.join(base, f'generated.out.{index}')

    print(f"Running Test {index}...")
    dfa = parse_dfa_from_csv(input_file)
    minimized = dfa.minimize()
    write_dfa_to_csv(minimized, generated_output)

    if compare_csv_files(expected_output, generated_output):
        print(f"✅ Test {index} passed.")
    else:
        print(f"❌ Test {index} failed. Compare {expected_output} vs {generated_output}.")


if __name__ == '__main__':
    # Execute tests 1–3 by default; adjust range as needed
    for i in range(1, 4):
        run_test(i)

File: projects/13-Hopcroft/implementation/Hopcroft.java
import java.io.*;
import java.nio.file.Files;
import java.util.*;

public class Hopcroft {

    public static void main(String[] args) {
        try {
            for (int i = 1; i <= 3; i++) {
                runTest(i);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static class DFA {
        Set<String> states;
        Set<String> alphabet;
        Map<TransitionKey, String> tf;
        String start;
        Set<String> accept;

        DFA(Set<String> states, Set<String> alphabet,
            Map<TransitionKey, String> tf, String start, Set<String> accept) {
            this.states = new HashSet<>(states);
            this.alphabet = new HashSet<>(alphabet);
            this.tf = new HashMap<>(tf);
            this.start = start;
            this.accept = new HashSet<>(accept);
        }

        void removeUnreachable() {
            Set<String> reachable = new HashSet<>();
            Deque<String> queue = new ArrayDeque<>();
            queue.add(start);
            while (!queue.isEmpty()) {
                String s = queue.poll();
                if (reachable.add(s)) {
                    for (String c : alphabet) {
                        String t = tf.get(new TransitionKey(s, c));
                        if (t != null && !reachable.contains(t)) {
                            queue.add(t);
                        }
                    }
                }
            }
            states.retainAll(reachable);
            accept.retainAll(reachable);
            tf.entrySet().removeIf(e -> {
                String s = e.getKey().state;
                String t = e.getValue();
                return !reachable.contains(s) || !reachable.contains(t);
            });
        }

        void makeTotal() {
            String trap = "__TRAP__";
            boolean needsTrap = false;
            for (String s : new HashSet<>(states)) {
                for (String c : alphabet) {
                    TransitionKey key = new TransitionKey(s, c);
                    if (!tf.containsKey(key)) {
                        tf.put(key, trap);
                        needsTrap = true;
                    }
                }
            }
            if (needsTrap) {
                states.add(trap);
                for (String c : alphabet) {
                    tf.put(new TransitionKey(trap, c), trap);
                }
            }
        }

        DFA minimize() {
            removeUnreachable();
            makeTotal();

            // Initial partition: accepting vs non-accepting
            List<Set<String>> P = new ArrayList<>();
            Set<String> nonAccept = new HashSet<>(states);
            nonAccept.removeAll(accept);
            if (!accept.isEmpty()) P.add(new HashSet<>(accept));
            if (!nonAccept.isEmpty()) P.add(nonAccept);

            Deque<Set<String>> W = new ArrayDeque<>(P);

            while (!W.isEmpty()) {
                Set<String> A = W.poll();
                for (String c : alphabet) {
                    Set<String> X = new HashSet<>();
                    for (String s : states) {
                        String t = tf.get(new TransitionKey(s, c));
                        if (t != null && A.contains(t)) {
                            X.add(s);
                        }
                    }
                    List<Set<String>> newP = new ArrayList<>();
                    for (Set<String> Y : P) {
                        Set<String> inter = new HashSet<>(Y);
                        inter.retainAll(X);
                        Set<String> diff = new HashSet<>(Y);
                        diff.removeAll(X);
                        if (!inter.isEmpty() && !diff.isEmpty()) {
                            newP.add(inter);
                            newP.add(diff);
                            if (W.remove(Y)) {
                                W.add(inter);
                                W.add(diff);
                            } else {
                                W.add(inter.size() <= diff.size() ? inter : diff);
                            }
                        } else {
                            newP.add(Y);
                        }
                    }
                    P = newP;
                }
            }

            // Map each final block to a numeric label 0..n-1, ordered by original state names
            List<Set<String>> sortedBlocks = new ArrayList<>(P);
            sortedBlocks.sort(Comparator.comparing(b -> Collections.min(b)));
            Map<Set<String>, String> blockIndex = new HashMap<>();
            for (int i = 0; i < sortedBlocks.size(); i++) {
                blockIndex.put(sortedBlocks.get(i), Integer.toString(i));
            }

            // Build representative map
            Map<String, String> rep = new HashMap<>();
            for (Set<String> block : sortedBlocks) {
                String idx = blockIndex.get(block);
                for (String s : block) {
                    rep.put(s, idx);
                }
            }

            // Construct new DFA
            Set<String> newStates = new HashSet<>(blockIndex.values());
            String newStart = rep.get(start);
            Set<String> newAccept = new HashSet<>();
            for (String s : accept) newAccept.add(rep.get(s));
            Map<TransitionKey, String> newTf = new HashMap<>();
            for (Set<String> block : sortedBlocks) {
                String any = Collections.min(block);
                String from = rep.get(any);
                for (String c : alphabet) {
                    String toState = tf.get(new TransitionKey(any, c));
                    if (toState != null) newTf.put(new TransitionKey(from, c), rep.get(toState));
                }
            }

            return new DFA(newStates, alphabet, newTf, newStart, newAccept);
        }
    }

    static class TransitionKey {
        String state;
        String symbol;

        TransitionKey(String state, String symbol) {
            this.state = state;
            this.symbol = symbol;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof TransitionKey)) return false;
            TransitionKey that = (TransitionKey) o;
            return state.equals(that.state) && symbol.equals(that.symbol);
        }

        @Override
        public int hashCode() {
            return Objects.hash(state, symbol);
        }
    }

    static DFA parseDFAFromCSV(String filename) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String[] fields = br.readLine().split(",");
            List<String> alphabet = Arrays.asList(fields).subList(3, fields.length);

            Set<String> states = new HashSet<>();
            Map<TransitionKey, String> tf = new HashMap<>();
            String start = null;
            Set<String> accept = new HashSet<>();

            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                String s = parts[0];
                states.add(s);
                if (parts[1].equals("1")) start = s;
                if (parts[2].equals("1")) accept.add(s);
                for (int i = 3; i < parts.length; i++) {
                    tf.put(new TransitionKey(s, alphabet.get(i - 3)), parts[i]);
                }
            }
            return new DFA(states, new HashSet<>(alphabet), tf, start, accept);
        }
    }

    static void writeDFAtoCSV(DFA dfa, String filename) throws IOException {
        List<String> sortedAlpha = new ArrayList<>(dfa.alphabet);
        Collections.sort(sortedAlpha);
        List<String> sortedStates = new ArrayList<>(dfa.states);
        sortedStates.sort(Comparator.comparingInt(Integer::parseInt));

        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filename))) {
            // Header
            bw.write("state,is_start,is_accept");
            for (String c : sortedAlpha) bw.write("," + c);
            bw.newLine();
            // Rows
            for (String s : sortedStates) {
                bw.write(s + "," + (s.equals(dfa.start) ? "1" : "0") + "," + (dfa.accept.contains(s) ? "1" : "0"));
                for (String c : sortedAlpha) {
                    String to = dfa.tf.get(new TransitionKey(s, c));
                    bw.write("," + (to != null ? to : ""));
                }
                bw.newLine();
            }
        }
    }

    static boolean compareCSVFiles(String f1, String f2) throws IOException {
        List<String> a = Files.readAllLines(new File(f1).toPath());
        List<String> b = Files.readAllLines(new File(f2).toPath());
        return a.equals(b);
    }

    static void runTest(int index) throws IOException {
        String base = "io" + File.separator;
        String in = base + "sample.in." + index;
        String out = base + "sample.out." + index;
        String gen = base + "generated.out." + index;

        System.out.println("Running Test " + index + "...");
        DFA dfa = parseDFAFromCSV(in);
        DFA min = dfa.minimize();
        writeDFAtoCSV(min, gen);
        if (compareCSVFiles(out, gen)) {
            System.out.println("✅ Test " + index + " passed.");
        } else {
            System.out.println("❌ Test " + index + " failed. Compare " + out + " and " + gen + ".");
        }
    }
}
